src/bottom_up_iterator.jl

```diff
@@ -12,235 +12,212 @@ Concrete implementations of this iterator should implement the following methods
 abstract type BottomUpIterator <: ProgramIterator end
 
 Base.@doc """
-    @programiterator BasicIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
+    @programiterator DepthIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
 
 A basic implementation of the bottom-up iterator. It will enumerate all programs in increasing order based on their depth.
-""" BasicIterator
-@programiterator BasicIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
+""" DepthIterator
+@programiterator DepthIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
 
-"""
-	struct BottomUpState
-
-Holds the state of the bottom-up iterator. This includes the priority bank, the hashes of the outputs of the programs, and the current programs that are being enumerated.
-"""
-struct BottomUpState
-    priority_bank::Dict{Symbol, Dict{RuleNode,Int64}}
-    hashes::Set{UInt}
-    current_programs::Queue{RuleNode}
+mutable struct BottomUpState
+    bank::Any
+    data::Any
+    hash_set::Set{UInt}
 end
 
-"""
-	order(iter::BottomUpIterator, grammar::ContextSensitiveGrammar)::Vector{Int64}
-
-Returns the order in which the rules should be enumerated.
-"""
-function order(
-    ::BottomUpIterator,
-    grammar::ContextSensitiveGrammar
-)
```

Comment:
add return type so that it is clear what should the function return (from the name, I did not expect an iterator; does it have to be?)